tests/mman/enc/enc.c

```diff
@@ -0,0 +1,317 @@
+    m = oe_test_get_mappings();
+    OE_TEST(m->start == p2_start);
+    OE_TEST(m->end == p2_end);
+    m = m->next;
+    OE_TEST(m->start == p1_start);
+    OE_TEST(m->end == p1_end);
+
+    // Do another partial unmap.
+    start -= OE_PAGE_SIZE;
+    OE_TEST(munmap((void*)start, end - start) == 0);
+    OE_TEST(errno == 0);
+    m = oe_test_get_mappings();
+    OE_TEST(m->start == p2_start);
+    OE_TEST(m->end == p2_end);
+    m = m->next;
+    OE_TEST(m->start == p1_start);
+    OE_TEST(m->end == p1_end);
+
+    // Do an unmap till the start.
+    // This ought to delete one mapping completely.
+    OE_TEST(munmap((void*)OE_PAGE_SIZE, start - OE_PAGE_SIZE) == 0);
+    OE_TEST(errno == 0);
+    m = oe_test_get_mappings();
+    OE_TEST(m->next == NULL);
+
+    // Do another unmapping that spans entire enclave memory.
+    // This ought to get rid of all mappings.
+    for (size_t i = 1; i < 20; ++i)
+    {
+        OE_TEST(
+            mmap(NULL, i * 1, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) !=
+            MAP_FAILED);
+        OE_TEST(errno == 0);
+    }
+    OE_TEST(munmap(0, (1L << 62)) == 0);
+    OE_TEST(errno == 0);
+    OE_TEST(oe_test_get_mappings() == NULL);
+
+    // Test unmapping a mapping in small chunks.
+    start = (uint64_t)mmap(
+        NULL, 3 * OE_PAGE_SIZE, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+    OE_TEST(oe_test_get_mappings() != NULL);
+
+    OE_TEST(munmap((void*)(start + OE_PAGE_SIZE), 1) == 0);
+    OE_TEST(oe_test_get_mappings() != NULL);
+    OE_TEST(munmap((void*)(start + 2 * OE_PAGE_SIZE), 1) == 0);
+    OE_TEST(oe_test_get_mappings() != NULL);
+    OE_TEST(munmap((void*)start, 1) == 0);
+    OE_TEST(oe_test_get_mappings() == NULL);
+}
+
+static void _test_mmap_params(void)
+{
+    // Non zero addr should fail.
+    OE_TEST(
+        mmap((void*)1, 1024, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) ==
+        MAP_FAILED);
+    OE_TEST(errno == EINVAL);
+
+    // Zero length should fail.
+    OE_TEST(
+        mmap(NULL, 0, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) ==
+        MAP_FAILED);
+    OE_TEST(errno == EINVAL);
+
+    // Large mmap  should fail.
+    OE_TEST(
+        mmap(NULL, (1L << 52), PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) ==
+        MAP_FAILED);
+    OE_TEST(errno == ENOMEM);
+
+    OE_TEST(
+        mmap(NULL, (1L << 32), PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) ==
+        MAP_FAILED);
+    OE_TEST(errno == ENOMEM);
+
+    // Test various prots.
+    OE_TEST(
+        mmap(NULL, 1024, 0, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) == MAP_FAILED);
+    OE_TEST(errno == EINVAL);
+
+    OE_TEST(
+        mmap(NULL, 1024, PROT_EXEC, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) ==
+        MAP_FAILED);
+    OE_TEST(errno == EINVAL);
+
+    OE_TEST(
+        mmap(NULL, 1024, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) !=
+        MAP_FAILED);
+    OE_TEST(errno == 0);
+
+    OE_TEST(
+        mmap(NULL, 5024, PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0) !=
```

Comment:
why 5024 instead of 1024 like the other lines?

Maybe define and use something like
```
   const size_t chunk_size = 1024;
```
at the top for use in all of these?

Reply:
Good idea. Done.
This particular test is just to check for a multiple of chunk size.